File: CoreLibrary/OfficialLocalizations/Japanese/GUI_Components/Controls/_Validation/PreMadeRules/Strings/MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule.ts
import type { InputtedValueValidation } from "@yamato-daiwa/frontend";
import { MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRuleLocalization__Japanese } from
    "./MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRuleLocalization.japanese";
import { isNotUndefined } from "@yamato-daiwa/es-extensions";


export class MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule implements InputtedValueValidation.Rule<string> {

  public static localization: MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule.Localization =
      MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRuleLocalization__Japanese;

  public readonly mustFinishValidationIfValueIsInvalid: boolean;

  private readonly MINIMAL_KANJI_OR_KANA_CHARACTERS_RATIO: number;
  private readonly isKanjiOrKanaCharactersRatioLessThanRequiredMinimum?: (
    compoundParameter: Readonly<{ targetString: string; minimalKanjiOrKanaCharactersRatio: number; }>
  ) => boolean;

  private readonly errorMessageBuilder: MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule.ErrorMessage.Builder;


  public constructor(
    compoundParameter:
        InputtedValueValidation.Rule.ConstructorParameter &
        Readonly<{
          minimalKanjiOrKanaCharactersRatio: number;
          isKanjiOrKanaCharactersRatioLessThanRequiredMinimum?: (
            compoundParameter: Readonly<{ targetString: string; minimalKanjiOrKanaCharactersRatio: number; }>
          ) => boolean;
          errorMessageBuilder?: MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule.ErrorMessage.Builder;
          errorMessage?: string;
        }>
  ) {

    this.mustFinishValidationIfValueIsInvalid = compoundParameter.mustFinishValidationIfValueIsInvalid ?? false;

    this.MINIMAL_KANJI_OR_KANA_CHARACTERS_RATIO = compoundParameter.minimalKanjiOrKanaCharactersRatio;
    this.isKanjiOrKanaCharactersRatioLessThanRequiredMinimum = compoundParameter.
        isKanjiOrKanaCharactersRatioLessThanRequiredMinimum;

    if (isNotUndefined(compoundParameter.errorMessageBuilder)) {
      this.errorMessageBuilder = compoundParameter.errorMessageBuilder;
    } else if (isNotUndefined(compoundParameter.errorMessage)) {
      /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions --
       * It was proved that "errorMessage" is non-undefined, and it will not change. */
      this.errorMessageBuilder = (): string => compoundParameter.errorMessage as string;
    } else {
      this.errorMessageBuilder = MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule.localization.
          errorMessageBuilder;
    }

  }

  public check(rawValue: string): InputtedValueValidation.Rule.CheckingResult {

    const isKanjiOrKanaCharactersRatioLessThanRequiredMinimum: boolean =
        this.isKanjiOrKanaCharactersRatioLessThanRequiredMinimum?.({
          targetString: rawValue,
          minimalKanjiOrKanaCharactersRatio: this.MINIMAL_KANJI_OR_KANA_CHARACTERS_RATIO
        }) ??
        ((): boolean => {

          let kanjiOrKanaCharactersCount: number = 0;

          for (const character of rawValue) {

            if ((/[ぁ-んァ-ヶｱ-ﾝﾞﾟ一-龠]/u).test(character)) {

              kanjiOrKanaCharactersCount++;

              const kanjiOrKanaCharactersRatio: number = kanjiOrKanaCharactersCount / rawValue.length;

              if (kanjiOrKanaCharactersRatio >= this.MINIMAL_KANJI_OR_KANA_CHARACTERS_RATIO) {
                return false;
              }

            }

          }

          return true;

        })();

    return isKanjiOrKanaCharactersRatioLessThanRequiredMinimum ?
        {
          isValid: false,
          errorMessage: this.errorMessageBuilder({
            rawValue, minimalKanjiOrKanaCharactersRatio: this.MINIMAL_KANJI_OR_KANA_CHARACTERS_RATIO
          })
        } :
        { isValid: true };

  }

}


export namespace MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule {

  export type Localization = Readonly<{ errorMessageBuilder: ErrorMessage.Builder; }>;

  export namespace ErrorMessage {

    export type Builder = (templateVariables: TemplateVariables) => string;

    export type TemplateVariables = Readonly<{
      minimalKanjiOrKanaCharactersRatio: number;
      rawValue: string;
    }>;

  }

}

File: AdaptationsToFrameworks/Blazor/Localizations/Japanese/GUI_Components/Controls/Validation/PreMadeRules/Strings/MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule.cs
using System.Text.RegularExpressions;
using YamatoDaiwa.Frontend.GUI_Components.Controls.Validation;

namespace YamatoDaiwa.Frontend.Localizations.Japanese.GUI_Components.Controls.Validation.PreMadeRules.Strings;


public class MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule : InputtedValueValidation.IRule
{

  /* ━━━ Localization ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public interface ILocalization
  {

    public Func<ErrorMessage.TemplateVariables, string> ErrorMessageBuilder { get; }

    public static class ErrorMessage
    {

      public record TemplateVariables
      {
        public required string RawValue { get; init; }
        public required IConvertible MinimalKanjiOrKanaCharactersRatio { get; init; }
      }

    }

  }

  public static ILocalization Localization { get; set; } = new MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRuleJapaneseLocalization();


  /* ━━━ Properties ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public bool MustFinishValidationIfValueIsInvalid { get; init; }

  public required IConvertible MinimalKanjiOrKanaCharactersRatio { get; init; }

  public record CustomCheckParameter
  {
    public required string TargetString { get; init; }
    public required double MinimalKanjiOrKanaCharactersRatio { get; init; }
  }

  public Func<CustomCheckParameter, bool>? IsKanjiOrKanaCharactersRatioLessThanRequiredMinimum { get; init; }

  public Func<ILocalization.ErrorMessage.TemplateVariables, string>? ErrorMessageBuilder { get; init; }
  public string? ErrorMessage { get; init; }

  private static readonly Regex KanjiOrKanaRegex = new(@"[ぁ-んァ-ヶｱ-ﾝﾞﾟ一-龠]", RegexOptions.Compiled);


  /* ━━━ Interface Implementation ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public InputtedValueValidation.IRule.CheckingResult Check(object rawValue)
  {

    double minimalKanjiOrKanaCharactersRatio = Convert.ToDouble(this.MinimalKanjiOrKanaCharactersRatio);

    bool isInvalid = this.IsKanjiOrKanaCharactersRatioLessThanRequiredMinimum?.
        Invoke(
          new CustomCheckParameter
          {
            TargetString = (string)rawValue,
            MinimalKanjiOrKanaCharactersRatio = minimalKanjiOrKanaCharactersRatio
          }
        ) ?? 
        checkByDefaultAlgorithm((string)rawValue, minimalKanjiOrKanaCharactersRatio);


    return new InputtedValueValidation.IRule.CheckingResult
    {
      ErrorMessage = isInvalid ?
          this.buildErrorMessage(
            new ILocalization.ErrorMessage.TemplateVariables
            {
              RawValue = (string)rawValue,
              MinimalKanjiOrKanaCharactersRatio = this.MinimalKanjiOrKanaCharactersRatio
            }
          ) :
          null
    };

  }

  private static bool checkByDefaultAlgorithm(string targetString, double minimalKanjiOrKanaCharactersRatio)
  {

    int kanjiOrKanaCharactersCount = 0;

    foreach (char character in targetString)
    {

      if (KanjiOrKanaRegex.IsMatch(character.ToString()))
      {

        kanjiOrKanaCharactersCount++;

        double kanjiOrKanaCharactersRatio = (double)kanjiOrKanaCharactersCount / targetString.Length;

        if (kanjiOrKanaCharactersRatio >= minimalKanjiOrKanaCharactersRatio)
        {
          return false;
        }

      }

    }

    return true;

  }

  private string buildErrorMessage(ILocalization.ErrorMessage.TemplateVariables templateVariables)
  {
    return this.ErrorMessageBuilder?.Invoke(templateVariables) ??
        this.ErrorMessage ??
        MinimalKanjiOrKanaCharactersRatioInputtedValueValidationRule.Localization.ErrorMessageBuilder(templateVariables);
  }

}
